Add symfony mailer adapter

pocky opened the pull request:
Add SymfonyMailerAdapter to SyliusMailer. This code is already in production on two projects.

pocky commented:
Travis failed with symfony/mailer:4.3 because `new Address()` take only one parameter and we should use `new NamedAddress()` in this version. `NamedAddress` object don't exist in symfony/mailer:^4.4|^5.0.

Should I ignore this error (and update travis/composer.json) because sylius 1.7 require symfony/*:^4.4 ?

lchrusciel commented:
Hello Alexandre,

thanks a lot for your contribution and welcome to our community! That's a great PR, we would love to support Symfony Mailer. Support for the 4.3 can be dropped, so it is not an issue.

However, this class even if speced is a dead code right now. We need to register it conditionally in container, so it can be easily used. What is more, we need to test that this bundle will work with Symfony mailer added to the package, and with both Swiftmailer and Symfony mailer added as well(in this case, it should fall back to the current behavior, so by default, we should use Swiftmailer). It can be tested similar to this PR: https://github.com/Sylius/SyliusMailerBundle/pull/13/files#diff-4ea6733e1e3efdbcae2ffa4c721646acR28-R105

It would be nice, to have maybe some functional tests as well, to see if it really works. The last (and the least) would be documentation for it, but first, we should move the current documentation to this package and convert it to markdown.

Would you be willing to do anything from my suggestions?

pocky commented:
@lchrusciel Hello! 
 
- Symfony 4.3 is dropped from travis
- Tests are updated

I changed the code of the `SyliusMailerExtension` because it's impossible to detect a component like a Bundle. Being based on a test like `interface_exists(MailerInterface::class)` is not viable because it will always be true if symfony/mailer is installed. So I modified the code to rely on the services defined by the container.

However, I have a problem with phpstan that I don't know how to solve.

For the documentation, my English is not good enough to write something, sorry :x

lchrusciel commented on src/Bundle/Sender/Adapter/SymfonyMailerAdapter.php:
```diff
@@ -0,0 +1,82 @@
+    /** @var EventDispatcherInterface|null */
+    protected $dispatcher;
+
+    public function __construct(MailerInterface $mailer, ?EventDispatcherInterface $dispatcher = null)
```
As this is a new class, we can force its dependency. 
```suggestion
    public function __construct(MailerInterface $mailer, EventDispatcherInterface $dispatcher)
```

pocky replied:
As you want but EventDispatcherInterface is also nullable on DefaultAdapter.

lchrusciel replied:
That's a good point, and should be fixed there as well. But that could be a topic for another PR.

lchrusciel commented on src/Bundle/Sender/Adapter/SymfonyMailerAdapter.php:
```diff
@@ -0,0 +1,82 @@
+    /** @var MailerInterface */
+    protected $mailer;
+
+    /** @var EventDispatcherInterface|null */
```
```suggestion
    /** @var EventDispatcherInterface*/
```

lchrusciel commented on src/Bundle/Sender/Adapter/SymfonyMailerAdapter.php:
```diff
@@ -0,0 +1,82 @@
+use Symfony\Component\Mime\Email;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+class SymfonyMailerAdapter extends AbstractAdapter
```
```suggestion
final class SymfonyMailerAdapter extends AbstractAdapter
```

lchrusciel commented on src/Bundle/test/src/Tests/DependencyInjection/SyliusMailerExtensionTest.php:
```diff
@@ -91,6 +92,55 @@ public function it_configures_twig_and_swiftmailer_adapters_if_they_are_availabl
+
+        $this->assertTrue(interface_exists(MailerInterface::class));
+        $this->assertContainerBuilderHasService('sylius.email_sender.adapter.symfony_mailer', SymfonyMailerAdapter::class);
+        $this->assertContainerBuilderHasAlias('sylius.email_sender.adapter', 'sylius.email_sender.adapter.symfony_mailer');
```
Seems wrong if I'm correct

pocky replied:
I have updated the function name.

I made this test because I want to be sure that SymfonyMailer will be used if SwiftMailer is also present in the project (because SwiftMailer is on Sylius/Sylius Standard and SymfonyMailer is installed by the developer).

lchrusciel replied:
IMHO, we should cover this other way around. Right now, one could add a Symfony mailer to their end Sylius application(to use it for other mailings). If we will merge it with the current state, the application may not work properly(due to some changes, as the default adapter has been switched. It may be considered BC Break. 

However, usage of SymfonyMailer by default may not be a bad option, if we will mention this change in UPGRADE file.

Zales0123 commented:
Closing in favor of https://github.com/Sylius/SyliusMailerBundle/pull/102, thank you, @pocky for your contribution 🖖 🎉
